inspirehep/modules/multieditor/actions.py

```diff
@@ -0,0 +1,348 @@
+#
+# In applying this license, CERN does not waive the privileges and immunities
+# granted to it by virtue of its status as an Intergovernmental Organization
+# or submit itself to any jurisdiction.
+
+from __future__ import absolute_import, print_function, division
+
+import re
+
+from dictdiffer import diff
+from jsonschema import ValidationError
+from inspire_schemas.api import validate
+from inspire_utils.helpers import force_list
+
+
+class ActionFactory(object):
+
+    @staticmethod
+    def factory(action_type, **kwargs):
+        if action_type == 'Addition':
+            return Addition(**kwargs)
+        elif action_type == 'Deletion':
+            return Deletion(**kwargs)
+        elif action_type == 'Update':
+            return Update(**kwargs)
+
+
+class Action(object):
+    def __init__(self, keypath, value=None, match_type=None, value_to_check=None,
+                 conditions=None):
+        """
+        :param keypath: path to value
+        :type keypath: array(string)
+
+       :param value: value to add or update to
+       :type value: string
+
+       :param match_type: type of matching
+       :type match_type: string
+
+       :param value_to_check: value to replace or delete
+       :type value_to_check: string
+
+       :param conditions: conditions of the action
+       :type conditions: array(object)
+        """
+        self.keypath = keypath
+        self.value = value
+        self.match_type = match_type
+        self.value_to_check = value_to_check
+        self.conditions = conditions
+        self.changed = False
+
+    def process_keypath(self, record, position, conditions_passed):
+        key = self.keypath[position]
+        condition_failed = False
+        if self.conditions:
+            for condition in self.conditions:
+                if position < len(condition.get('keypath')) and\
+                        (key != condition.get('keypath')[position] and
+                            (position == 0 or self.keypath[position - 1] == condition.get('keypath')[position - 1]) or
+                            (key == condition.get('keypath')[position] and position == len(self.keypath) - 1)):
+                    # Condition and action key should differ for the first time to be checked
+                    # Or action key is the last on action's path and condition path is the same or extends deeper
+                    if not check_value(record=record, keypath=condition.get('keypath'),
+                                       value_to_check=condition.get('value', ''),
+                                       match_type=condition.get('match_type', ''), position=position):
+                        condition_failed = True
+                    conditions_passed = conditions_passed + 1  # number of conditions that passed successfully
+        return conditions_passed, key, condition_failed
+
+
+class Addition(Action):
+
+    def apply(self, record, schema, position=0, conditions_passed=0):
```

Comment:
since we have class maybe it's ok store these states  (`position` and `condition_passed`) as an instance property instead of passing everytime.